# Q6 Dice Simulation (Monte Carlo).

# Prompt
Create roll() -> int that stimulates a fair die, with 100,000 rolls to estimate the probability that sum of two dice equals 7. compare empirical result to the exact probability.
# Chatgpt Code

In [1]:
import random

def roll() -> int:
    return random.randint(1, 6)

def estimate_probability(target_sum: int = 7, trials: int = 100_000) -> None:
    count = 0
    for _ in range(trials):
        if roll() + roll() == target_sum:
            count += 1

    empirical = count / trials
    exact = 6 / 36  # There are 6 combinations that sum to 7 out of 36 total
    print(f"Empirical probability of sum == {target_sum}: {empirical:.4f}")
    print(f"Exact probability of sum == {target_sum}: {exact:.4f}")
    print(f"Difference: {abs(empirical - exact):.4f}")

if __name__ == "__main__":
    estimate_probability()

Empirical probability of sum == 7: 0.1673
Exact probability of sum == 7: 0.1667
Difference: 0.0007


# Critqiue

-Correctness: The program correctly simulates a fair die using random.randint(1,6) and performs 100,000 trials to estimate the probability that the sum of two dice equals 7. It accurately compares the empirical result to the theoretical probability of ( \frac{6}{36} = 0.1667 ), which reflects the six favorable outcomes out of 36 total combinations.

-Complexity: The simulation runs in linear time—O(n)— where n is the number of trials. Each iteration performs two constant-time operations (rolling two dice), making it efficient even for large sample sizes like 100,000.

-Robustness: The code handles the simulation reliably but assumes ideal conditions (e.g., no input errors, fixed trial count). It could be improved by allowing user-defined trial counts, validating inputs, or seeding the random generator for reproducibility in testing scenarios.

-Readability: The function is well-structured and easy to follow. Variable names like count, trials, and target_sum are clear, and the output is formatted for quick comparison. 

-Faithfullness: The implementation stays true to its goal: simulating fair dice rolls and estimating a specific probability. It uses standard Python tools appropriately and avoids unnecessary complexity, making it a faithful and educational representation of empirical probability testing.

# Improved Code

In [2]:
import random

def roll():
    return random.randint(1, 6)

count = 0
trials = 100_000

for _ in range(trials):
    if roll() + roll() == 7:
        count += 1

empirical = count / trials
exact = 6 / 36

print(f"Empirical probability: {empirical:.4f}")
print(f"Exact probability:     {exact:.4f}")
print(f"Difference:            {abs(empirical - exact):.4f}")

Empirical probability: 0.1687
Exact probability:     0.1667
Difference:            0.0020


Use random.seed() when you need repeatable results—especially in simulations, testing, or academic work. Omit it when you want true randomness across runs, like in games or cryptographic applications.